### PR TITLE
Rework bitstring ops using templates instead of macros

### PR DESCRIPTION
**Pull request by caleridas**
Remove all code generation macros for instantiating bitstring operation classes. Replace this with normal C++ template classes.

### CODE REVIEW

File: jlm/rvsdg/bitstring/arithmetic.hpp [hunk 1]
```diff
@@ -12,91 +12,193 @@
...
-DECLARE_BITBINARY_OPERATION(bitshr)
-DECLARE_BITBINARY_OPERATION(bitsmod)
-DECLARE_BITBINARY_OPERATION(bitsmulh)
-DECLARE_BITBINARY_OPERATION(bitsub)
-DECLARE_BITBINARY_OPERATION(bitudiv)
-DECLARE_BITBINARY_OPERATION(bitumod)
-DECLARE_BITBINARY_OPERATION(bitumulh)
-DECLARE_BITBINARY_OPERATION(bitxor)
+template<typename reduction, const char * name>
+class make_bitunary_op final : public bitunary_op
```
Comment (phate) [inline jlm/rvsdg/bitstring/arithmetic.hpp:16]:
```
template<typename TReduction, const char * TName>
class MakeBitunaryOperation final : public bitunary_op
```

File: jlm/rvsdg/bitstring/arithmetic.hpp [hunk 2]
```diff
@@ -12,91 +12,193 @@
...
+{
+public:
+  ~make_bitunary_op() noexcept override;
+
+  inline make_bitunary_op(const bittype & type) noexcept
+      : bitunary_op(type)
+  {}
```
Comment (phate) [inline jlm/rvsdg/bitstring/arithmetic.hpp:23]: 1. Remove `inline`
2. Add `explicit`

File: jlm/rvsdg/bitstring/arithmetic.hpp [hunk 3]
```diff
@@ -12,91 +12,193 @@
...
+
+  virtual bool
+  operator==(const operation & other) const noexcept override;
```
Comment (phate) [inline jlm/rvsdg/bitstring/arithmetic.hpp:26]: `virtual` is redundant

File: jlm/rvsdg/bitstring/arithmetic.hpp [hunk 4]
```diff
@@ -12,91 +12,193 @@
...
+
+  virtual bitvalue_repr
+  reduce_constant(const bitvalue_repr & arg) const override;
```
Comment (phate) [inline jlm/rvsdg/bitstring/arithmetic.hpp:29]: `virtual` is redundant

File: jlm/rvsdg/bitstring/arithmetic.hpp [hunk 5]
```diff
@@ -12,91 +12,193 @@
...
+
+  virtual std::string
+  debug_string() const override;
```
Comment (phate) [inline jlm/rvsdg/bitstring/arithmetic.hpp:32]: `virtual` is redundant

File: jlm/rvsdg/bitstring/arithmetic.hpp [hunk 6]
```diff
@@ -12,91 +12,193 @@
...
+
+  virtual std::unique_ptr<operation>
+  copy() const override;
```
Comment (phate) [inline jlm/rvsdg/bitstring/arithmetic.hpp:35]: `virtual` is redundant

File: jlm/rvsdg/bitstring/arithmetic.hpp [hunk 7]
```diff
@@ -12,91 +12,193 @@
...
+
+  virtual std::unique_ptr<bitunary_op>
+  create(size_t nbits) const override;
```
Comment (phate) [inline jlm/rvsdg/bitstring/arithmetic.hpp:38]: `virtual` is redundant

File: jlm/rvsdg/bitstring/arithmetic.hpp [hunk 8]
```diff
@@ -12,91 +12,193 @@
...
+
+  static inline output *
+  create(size_t nbits, output * op)
```
Comment (phate) [inline jlm/rvsdg/bitstring/arithmetic.hpp:41]: Remove `inline`

File: jlm/rvsdg/bitstring/arithmetic.hpp [hunk 9]
```diff
@@ -12,91 +12,193 @@
...
+  {
+    return simple_node::create_normalized(op->region(), make_bitunary_op(nbits), { op })[0];
+  }
+};
+
+template<typename reduction, const char * name, enum binary_op::flags opflags>
+class make_bitbinary_op final : public bitbinary_op
```
Comment (phate) [inline jlm/rvsdg/bitstring/arithmetic.hpp:48]:
```
template<typename TReduction, const char * TName, enum binary_op::flags TOperationFlags>
class MakeBitbinaryOperation final : public bitbinary_op
```

File: jlm/rvsdg/bitstring/arithmetic.hpp [hunk 10]
```diff
@@ -12,91 +12,193 @@
...
+{
+public:
+  ~make_bitbinary_op() noexcept override;
+
+  inline make_bitbinary_op(const bittype & type) noexcept
+      : bitbinary_op(type)
+  {}
```
Comment (phate) [inline jlm/rvsdg/bitstring/arithmetic.hpp:55]: 1. Remove `inline`
2. Add `explicit`

File: jlm/rvsdg/bitstring/arithmetic.hpp [hunk 11]
```diff
@@ -12,91 +12,193 @@
...
+
+  virtual bool
+  operator==(const operation & other) const noexcept override;
```
Comment (phate) [inline jlm/rvsdg/bitstring/arithmetic.hpp:58]: `virtual` is redundant

File: jlm/rvsdg/bitstring/arithmetic.hpp [hunk 12]
```diff
@@ -12,91 +12,193 @@
...
+
+  virtual enum binary_op::flags
+  flags() const noexcept override;
```
Comment (phate) [inline jlm/rvsdg/bitstring/arithmetic.hpp:61]: `virtual` is redundant

File: jlm/rvsdg/bitstring/arithmetic.hpp [hunk 13]
```diff
@@ -12,91 +12,193 @@
...
+
+  virtual bitvalue_repr
+  reduce_constants(const bitvalue_repr & arg1, const bitvalue_repr & arg2) const override;
```
Comment (phate) [inline jlm/rvsdg/bitstring/arithmetic.hpp:64]: `virtual` is redundant

File: jlm/rvsdg/bitstring/arithmetic.hpp [hunk 14]
```diff
@@ -12,91 +12,193 @@
...
+
+  virtual std::string
+  debug_string() const override;
```
Comment (phate) [inline jlm/rvsdg/bitstring/arithmetic.hpp:67]: `virtual` is redundant

File: jlm/rvsdg/bitstring/arithmetic.hpp [hunk 15]
```diff
@@ -12,91 +12,193 @@
...
+
+  virtual std::unique_ptr<operation>
+  copy() const override;
```
Comment (phate) [inline jlm/rvsdg/bitstring/arithmetic.hpp:70]: `virtual` is redundant

File: jlm/rvsdg/bitstring/arithmetic.hpp [hunk 16]
```diff
@@ -12,91 +12,193 @@
...
+
+  virtual std::unique_ptr<bitbinary_op>
+  create(size_t nbits) const override;
```
Comment (phate) [inline jlm/rvsdg/bitstring/arithmetic.hpp:73]: `virtual` is redundant

File: jlm/rvsdg/bitstring/arithmetic.hpp [hunk 17]
```diff
@@ -12,91 +12,193 @@
...
+
+  static inline output *
+  create(size_t nbits, output * op1, output * op2)
```
Comment (phate) [inline jlm/rvsdg/bitstring/arithmetic.hpp:76]: Remove `inline`

File: jlm/rvsdg/Makefile.sub [hunk 18]
```diff
@@ -36,6 +36,7 @@ librvsdg_HEADERS = \
 	jlm/rvsdg/notifiers.hpp \
 	jlm/rvsdg/bitstring/constant.hpp \
 	jlm/rvsdg/bitstring/slice.hpp \
+	jlm/rvsdg/bitstring/arithmetic-impl.hpp \
```
Comment (phate) [inline jlm/rvsdg/Makefile.sub:39]: Could we align with the new naming convention here, calling it `ArithmeticImplementation.hpp` or so.

File: jlm/rvsdg/bitstring/arithmetic.cpp [hunk 19]
```diff
@@ -1,115 +1,240 @@
...
+  {
+    return arg.lnot();
+  }
+};
+
+const char BITNEGATE_LABEL[] = "BITNEGATE";
+template class make_bitunary_op<reduce_neg, BITNEGATE_LABEL>;
+
+const char BITNOT_LABEL[] = "BITNOT";
+template class make_bitunary_op<reduce_not, BITNOT_LABEL>;
```
Comment (phate) [inline jlm/rvsdg/bitstring/arithmetic.cpp:43]: Could we group methods/structs that belong to a single bit operation together? Right now, you have all the reductions first, then comes the template class definitions. I would rather have:

1. Everything belonging to bitnegate
2. Everything belonging to bitnot
3. ....

You did it nicely in `arithmetic.hpp`. Why not here?

File: jlm/rvsdg/bitstring/arithmetic.cpp [hunk 20]
```diff
@@ -1,115 +1,240 @@
...
+struct reduce_not
+{
+  bitvalue_repr
+  operator()(const bitvalue_repr & arg) const
+  {
+    return arg.lnot();
+  }
+};
+
+const char BITNEGATE_LABEL[] = "BITNEGATE";
```
Comment (phate) [inline jlm/rvsdg/bitstring/arithmetic.cpp:39]: I think we can also move away from the capital letters as debug strings. `BitNegate` would just do nicely as well.

**Review by phate — CHANGES_REQUESTED** [review]
The biggest "issue" I have with this PR is cosmetics and naming. We try to move over to a PascalCase naming convention, but it is not enforced yet. I would love if you could follow it in the newly introduced names (such that the PR stays contained to already touched files). It should be mostly `*_LABEL` and `reduce_*` that needs renaming.

The other comments I left is all cosmetics. However, since you are already touching those lines anyway, we could also fix it.

File: jlm/rvsdg/bitstring/arithmetic.hpp (hunk 1 above)
Comment (caleridas) [reply jlm/rvsdg/bitstring/arithmetic.hpp]: fixed

File: jlm/rvsdg/bitstring/arithmetic.hpp (hunk 3 above)
Comment (caleridas) [reply jlm/rvsdg/bitstring/arithmetic.hpp]: fixed

File: jlm/rvsdg/bitstring/arithmetic.hpp (hunk 17 above)
Comment (caleridas) [reply jlm/rvsdg/bitstring/arithmetic.hpp]: I know that inline is implicit when stated in class declaration, although I generally prefer to keep it to make it clear that I did not put it into the class decl by accident but by intent of making it inline. Removed it for now anyways.

File: jlm/rvsdg/Makefile.sub (hunk 18 above)
Comment (caleridas) [reply jlm/rvsdg/Makefile.sub:39]: Would it not be better to bulk-rename all files later?

File: jlm/rvsdg/bitstring/arithmetic.cpp (hunk 19 above)
Comment (caleridas) [reply jlm/rvsdg/bitstring/arithmetic.cpp]: done

File: jlm/rvsdg/bitstring/arithmetic.cpp (hunk 20 above)
Comment (caleridas) [reply jlm/rvsdg/bitstring/arithmetic.cpp]: Renamed, and also renamed the constant names themselves -- not sure if you did mean that as well?

File: jlm/rvsdg/Makefile.sub (hunk 18 above)
Comment (phate) [reply jlm/rvsdg/Makefile.sub:39]: Sure, let's leave it as it is for now.

**Review by phate — APPROVED** [review]

**Pull request merged**